native/src/main/java/io/ballerina/stdlib/xmldata/XmlToJson.java

```diff
@@ -202,13 +202,15 @@ private static void processAttributeWithAnnotation(BXmlItem xmlItem, String attr
                 } else if (fieldType instanceof UnionType) {
                     UnionType bUnionType = (UnionType) fieldType;
                     for (Type memberType : bUnionType.getMemberTypes()) {
-                        if (memberType.getTag() == TypeTags.RECORD_TYPE_TAG)  {
+                        Type referMemberType = TypeUtils.getReferredType(memberType);
+                        if (referMemberType.getTag() == TypeTags.RECORD_TYPE_TAG)  {
                             annotationType = memberType;
```

Comment:
```suggestion
                        Type referMemberType = TypeUtils.getReferredType(memberType);
                        if (referMemberType.getTag() == TypeTags.RECORD_TYPE_TAG)  {
                            annotationType = referMemberType;
```
I think, if we do like this, we can get rid of the changes in lines `211-213`